<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>CASSI: Genome-Wide Interaction Analysis Software</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<link rel="stylesheet" type="text/css" href="cassistyles.css" />
<link rel="shortcut icon" href="favicon.ico" />
</head>
<body>
<!-- Begin Wrapper -->
<div id="wrapper">
<table width="100%" border="0" cellpadding="0" cellspacing="0"><tr><td>
<!-- Begin Header -->
<div id="header">
<table><tr>
<td>
<img width="150" src="cassi-logo.png" alt="CASSI: Genome-Wide Interaction Analysis Software"/>
</td>
<td>
<p style="margin-left: 1em;">CASSI: Genome-Wide Interaction Analysis Software</p>
</td></tr>
</table>
</div>
<!-- End Header -->
<!-- Begin Menu Navigation -->
<ul id="menunav">
<li><a href="index.html">Home</a></li>
<li><a href="downloads.html">Download</a></li>
<li><a href="contact.html">Contact</a></li>
</ul>
<!-- End Menu Navigation -->

</td></tr>
<tr><td>
<table border="0" cellpadding="0" cellspacing="0"><tr>
<td valign="top">
<!-- Begin Left Column -->
<div id="leftcolumn">
<!-- Begin Contents Navigation -->
<div id="navcontainer">
<ul>
<li><a href="introduction.html">1 Introduction</a>
</li>
<li><a href="installation.html">2 Installation</a>
</li>
<li><a href="using.html">3 Using CASSI</a>
<ul>
<li><a href="using.html#input">3.1 Input Files</a></li>
<li><a href="using.html#basic-options">3.2 Options</a></li>
<li><a href="using.html#parameterfile">3.3 Parameter file</a></li>
<li><a href="using.html#memory">3.4 Memory options</a></li>
<li><a href="using.html#rsq">3.5 <i>R</i><sup>2</sup> and D'</a></li>
</ul>
</li>
<li><a href="joint-effects.html">4 Joint Effects</a>
<ul>
<li><a href="joint-effects.html#je-options">4.1 Options</a></li>
<li><a href="joint-effects.html#je-output">4.2 Output File</a></li>
</ul>
</li>
<li><a href="adjusted-wu.html">5 Adjusted Wu</a>
<ul>
<li><a href="adjusted-wu.html#awu-options">5.1 Options</a></li>
<li><a href="adjusted-wu.html#awu-output">5.2 Output File</a></li>
</ul>
</li>
<li><a href="adjusted-fast-epistasis.html">6 Adjusted Fast Epistasis</a>
<ul>
<li><a href="adjusted-fast-epistasis.html#afe-options">6.1 Options</a></li>
<li><a href="adjusted-fast-epistasis.html#afe-output">6.2 Output File</a></li>
</ul>
</li>
<li><a href="wellek-ziegler.html">7 Wellek Ziegler</a>
<ul>
<li><a href="wellek-ziegler.html#wz-options">7.1 Options</a></li>
<li><a href="wellek-ziegler.html#wz-output">7.2 Output File</a></li>
</ul>
</li>
<li><a href="logistic-regression.html">8 Logistic Regression</a>
<ul>
<li><a href="logistic-regression.html#lr-options">8.1 Options</a></li>
<li><a href="logistic-regression.html#lr-output">8.2 Output File</a></li>
<li><a href="logistic-regression.html#covariates">8.3 Covariates</a></li>
</ul>
</li>
<li><a href="linear-regression.html">9 Linear Regression</a>
<ul>
<li><a href="linear-regression.html#lin-options">9.1 Options</a></li>
<li><a href="linear-regression.html#lin-output">9.2 Output File</a></li>
<li><a href="linear-regression.html#lin-covariates">9.3 Covariates</a></li>
</ul>
</li>
<li><a href="multiple-tests.html">10 Filtering Multiple Tests</a>
<ul>
<li><a href="multiple-tests.html#screening">10.1 Screening Tests</a></li>
<li><a href="multiple-tests.html#calculate-all">10.2 Calculating All Tests</a></li>
<li><a href="multiple-tests.html#run-times">10.3 Runtimes</a></li>
</ul>
</li>
<li><a href="references.html">References</a></li>
</ul>
</div>
<!-- End Contents Navigation -->
</div>
<!-- End Left Column -->
</td>
<td valign="top">
<!-- Begin Right Column -->
<div id="rightcolumn">
<h1>10 Filtering Multiple Tests</h1>
<p>
It is possible to simultaneously calculate more than one test in CASSI and to use the tests as screening steps if desired. </p>
<h2><a name="screening">10.1 Screening Tests</a></h2>
<p>
To use one test to screen for another one is easy with CASSI. There will be quite few options to give to CASSI, so the easiest thing to do will is to write a parameter file. For example, save the following as <tt>myparas.pf</tt> </p>
<pre>#Set up input and output files
-i chromosome1.bed
-o myresults.dat

#Do logistic regression test as a screening step
-lr
-lr-th 0.001

#Do logistic regression test with 4 covariates
-lr
-lr-th 0.0001
-lr-covar covariates.dat
</pre>
<p>
and then run CASSI as follows: </p>
<pre>./cassi -pf myparas.pf
</pre>
<p>
The above example uses logistic regression with a p-value threshold of 0.001 as a screening step for logistic regression with 4 covariates. The tests are ran in the order they appear in the parameter file or the command line. The output will look something like: </p>
<pre>SNP1 CHR1 ID1 BP1 SNP2 CHR2 ID2 BP2 LR_LOG_OR LR_CHISQ LR_P LR_COVAR_LOG_OR LR_COVAR_CHISQ LR_COVAR_P
1 1 rs3825075 207140 1022 1 rs11039351 5842422 0.729394 15.2732 9.30269e-05 0.730193 15.2866 9.23706e-05
1 1 rs3825075 207140 1188 1 rs12794064 6674133 0.753422 14.7342 0.00012378 0.766574 15.2141 9.59825e-05
1 1 rs3825075 207140 1355 1 rs71037479 7211075 -0.56288 18.998 1.30857e-05 -0.567815 19.282 1.12763e-05
13 1 rs659804 244010 2703 1 rs10831905 12733043 2.43204 17.5586 2.78587e-05 2.45784 17.9211 2.30255e-05
13 1 rs659804 244010 2717 1 rs10500759 12779164 2.05542 15.5899 7.86735e-05 2.07359 15.9522 6.49639e-05
</pre>
<p>
My test data of 3000 SNPs took about 7 minutes to calculate the 4,498,500 SNP pair tests, compared to about 22 hours without the screening step - and resulted in the same results. Nice. To speed things up further we can add another screening stage using the adjusted fast epistasis test: </p>
<pre>#Set up input and output files
-i chromosome1.bed
-o myresults.dat

#Do the adjusted fast epistasis test as a screening step
-afe
-afe-th 0.2

#Do logistic regression test as a screening step
-lr
-lr-th 0.001

#Do logistic regression test with 4 covariates
-lr
-lr-th 0.0001
-lr-covar covariates.dat
</pre>
<p>
The runtime for the above analysis took about 5 and a half minutes to analysis the same data set and resulted in the same results. Obviously, the smaller the p-value threshold that is set for a screening step the quicker the analysis will run, but setting it too small may result in missing some significant test results in the last test. A comparison of the runtimes of each test are shown in <a href="multiple-tests.html#run-times">section 10.3</a> which may act as a guide as to which order to run the tests. The log file will show how many tests passed at each step and the log file for the above analysis looked like this: </p>
<pre>CASSI: SNP interaction analysis software, v2.00
-----------------------------------------------
Copyright 2013 Richard Howey, GNU General Public License, v3
Institute of Genetic Medicine, Newcastle University

Parameters:
Input file: chromosome1.bed
Output file: myresults.dat
Log file: myresults.log
Start SNP of first SNP window: 1
End SNP of first window is the last SNP
Start SNP of second SNP window: 1
End SNP of second window is the last SNP
Maximum no. of results: 1000000
Filter: all statistic thresholds must pass

Data Summary Statistics:
Number of SNPs: 3000
Number of subjects: 2000
Number of cases: 1000 (50%)
Number of controls: 1000 (50%)
Number of missing: 0

Test Statistic: Adjusted Fast Epistasis
P-value threshold for case/control results: 0.2
P-value threshold for case only results: 0.2
Total SNP pairs calculated: 4498500
Total SNP pair statistics passing threshold: 1753516

Test Statistic: Logistic Regression
P-value threshold for case/control results: 0.001
Total SNP pairs calculated: 1753516
Total SNP pair statistics passing threshold: 4055

Test Statistic: Logistic Regression
Covariate file: covariates.dat (with missing value -9)
P-value threshold for case/control results: 0.0001
Total SNP pairs calculated: 4055
Total SNP pair statistics passing threshold: 461

Number of SNP pairs with results: 461

Run time: 5 minutes and 24 seconds
</pre>
<p>
It can be seen that the first step reduces the number of tests to be performed by logistic regression by about 60 percent. The final step only tests about 0.09 percent of the original SNP pairs, speeding up the analysis considerably. </p>
<h2><a name="calculate-all">10.2 Calculating All Tests</a></h2>
<p>
It is also possible to calculate a number of tests and report results whenever any of the p-value thresholds are passed. This is done with the <tt>-filter-any</tt> option. For example, using the following parameter file: </p>
<pre>#Set up input and output files
-i chromosome1.bed
-o myresults.dat

#Set filter to allow any p-value threshold to be passed
-filter-any

#Do the adjusted fast epistasis test
-afe
-afe-th 1e-4

#Do Wellek Ziegler test
-wz
-wz-th 1e-5
</pre>
<p>
will produce results where either the adjusted fast epistasis test gives a p-value less than 0.0001 or the Wellek Ziegler test gives a p-value less than 0.00001. </p>
<p>
Every test will be evaluated when using the <tt>-filter-any</tt> option so the order of the tests in this case is not so important. </p>
<h2><a name="run-times">10.3 Runtimes</a></h2>
<p>
Figure 1 shows the relative runtimes for each of the tests in CASSI as well as the fast epistasis test in PLINK and the logistic regression epistasis test in PLINK. There is not too much difference in speed between the Wellek Ziegler (WZ) test, the adjusted fast epistatis (AFE) test, the adjusted Wu (AWU) test, the joint effects (JE) test and the fast epistasis (FE) test in PLINK. The AFE test is about 1.3 times quicker than the FE test in PLINK and 2.8 times quicker than logistic regression (LR). LR in CASSI is considerably faster than LR in PLINK, being about 25 times faster and is about 220 times faster than LR with 4 covariates. The AFE test is about 660 times faster than LR with 4 covariates. </p>
<p>
The relative speeds of the test should act as a guide as to which order they should be calculated when calculating multiple tests at once. The best order will also depend on the thresholds used for each test and possibly the data. </p>
<div id="fig"><a href="runtimesCASSI2.png"><img src="runtimesCASSI2.png" border="0" class="figimg" width="700"></a><br />
<br />
Figure 1. Runtimes to calculate 1 million SNP pair tests for the different tests in CASSI, plus two tests from PLINK, based on 4,498,500 SNP pair calculations of all the SNP pairs between 3000 SNPs with 1000 cases and 1000 controls (using a threshold of 0.0001 to output results). The tests left to right are: Wellek Ziegler (WZ); adjusted fast epistatis (AFE); adjusted Wu (AWU); joint effects (JE); PLINK fast epistasis (PLINK FE); logistic regression (LR); linear regression (LIN); linear regression with 1 and 4 covariates (LIN+1 cov and LIN+4 covs); PLINK logistic regression (PLINK LR); logistic regression with 1 and 4 covariates (LR+1 cov and LR+4 covs). Timings are based on the 64 bit Linux versions (machine: 6-Core AMD Opteron TM Processor with 2.6 GHz CPUs).
</div>
<br />
<div class="prevnext"><span class="left"><a href="linear-regression.html">&lt;-prev</a>
</span></div><!-- End Right Column -->
</div>
</td>
</tr></table>
<!-- End Wrapper -->
</td></tr></table>
</div>
 <!-- Start of StatCounter Code for Default Guide --> <script type="text/javascript"> var sc_project=8085233; var sc_invisible=1; var sc_security="e4dbb33e"; </script> <script type="text/javascript" src="http://www.statcounter.com/counter/counter.js"></script> <noscript><div class="statcounter"><a title="tumblr visit counter" href="http://statcounter.com/tumblr/" target="_blank"><img class="statcounter" src="http://c.statcounter.com/8085233/0/e4dbb33e/1/" alt="tumblr visit counter"></a></div></noscript> <!-- End of StatCounter Code for Default Guide --></body>
</html>
